dask_match/expr/core.py

```diff
@@ -0,0 +1,578 @@
+import functools
```

Comment:
Maybe de-nest the `expr/` directory?  It seems like this is most of the real code, while the current `core.py` is actually more of the non-core element, wrapping API around.  I would probably flatten everything, and rename core.py to something else.  I'm not sure what exactly.  `api.py`?  Something better would be welcome.  

In general though, keeping things more flat would make me happy.

Reply:
Yeah, I'm okay with de-nesting now - I first expected there to be more "api' code, and thought we *may* want to leave space for "array" expressions. However, I now agree that it would be simpler to use something like `api.py`, `frame.py`, or `collection.py` for collection code.

Reply:
Called it `collection.py` for now, but this can be changed pretty easily if we decide on something else.